### ISSUE TITLE
GRPO: For the same batch of data, there is a significant difference in model performance between using trainer. predict() and using trainer. train().

### ISSUE THREAD

**Issue by yzhdut** [huggingface/trl#2937]
I found that when I started training, the model performed poorly and the reward remained the lowest, so I doubted if it could be trained successfully. But when I used trainer. predict() to test the same batch of data, I found that the effect was much better than when training. I'm confused, is this normal?

**Comment by qgallouedec** [comment]
Thanks for reporting, please provide a refrence code and your system info

**Comment by yzhdut** [comment]
> Thanks for reporting, please provide a refrence code and your system info

ok 
system info：
gpu：
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.40.07              Driver Version: 550.40.07      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 4090        Off |   00000000:01:00.0 Off |                  Off |
| 44%   79C    P2            434W /  450W |   22731MiB /  24564MiB |    100%      Default |


`Package                       Version
----------------------------- ------------
accelerate                    1.4.0
aiofiles                      23.2.1
aiohttp                       3.9.1
aiosignal                     1.3.1
altair                        5.2.0
annotated-types               0.6.0
antlr4-python3-runtime        4.9.3
anyio                         4.2.0
async-timeout                 4.0.3
attrs                         23.1.0
auto_gptq                     0.7.1
bitsandbytes                  0.45.2
blinker                       1.7.0
cachetools                    5.3.2
certifi                       2023.11.17
charset-normalizer            3.3.2
click                         8.1.7
cmake                         3.27.9
colorama                      0.4.6
contourpy                     1.2.0
cpm-kernels                   1.0.11
cycler                        0.12.1
datasets                      3.3.1
deepspeed                     0.12.4
dill                          0.3.7
docstring-parser              0.15
et-xmlfile                    1.1.0
exceptiongroup                1.2.0
fastapi                       0.108.0
ffmpy                         0.3.1
filelock                      3.13.1
fonttools                     4.47.0
frozenlist                    1.4.1
fsspec                        2023.10.0
gekko                         1.2.1
gitdb                         4.0.11
GitPython                     3.1.40
gradio                        3.38.0
gradio_client                 0.8.0
h11                           0.14.0
hjson                         3.1.0
httpcore                      1.0.2
httpx                         0.26.0
huggingface-hub               0.29.0
idna                          3.6
importlib-metadata            6.11.0
importlib-resources           6.1.1
jieba                         0.42.1
Jinja2                        3.1.2
joblib                        1.3.2
jsonschema                    4.20.0
jsonschema-specifications     2023.11.2
kiwisolver                    1.4.5
linkify-it-py                 2.0.3
lit                           17.0.6
markdown-it-py                2.2.0
MarkupSafe                    2.1.3
matplotlib                    3.8.2
mdit-py-plugins               0.3.3
mdurl                         0.1.2
mpmath                        1.3.0
multidict                     6.0.4
multiprocess                  0.70.15
networkx                      3.2.1
ninja                         1.11.1.1
nltk                          3.8.1
numpy                         1.26.2
omegaconf                     2.3.0
openpyxl                      3.1.2
optimum                       1.24.0
orjson                        3.9.10
packaging                     23.2
pandas                        2.1.4
peft                          0.10.0
Pillow                        10.1.0
pip                           25.0.1
protobuf                      4.25.1
psutil                        5.9.6
py-cpuinfo                    9.0.0
pyarrow                       19.0.1
pyarrow-hotfix                0.6
pydantic                      2.5.2
pydantic_core                 2.14.5
pydeck                        0.8.1b0
pydub                         0.25.1
Pygments                      2.17.2
pynvml                        11.5.0
pyparsing                     3.1.1
pyproject                     1.3.1
python-dateutil               2.8.2
python-multipart              0.0.6
pytz                          2023.3.post1
PyYAML                        6.0.1
referencing                   0.32.0
regex                         2023.10.3
requests                      2.32.3
rich                          13.7.0
rouge                         1.0.1
rouge-chinese                 1.0.3
rpds-py                       0.13.2
safetensors                   0.5.2
scipy                         1.11.4
semantic-version              2.10.0
sentencepiece                 0.1.99
setuptools                    75.8.0
shellingham                   1.5.4
shtab                         1.6.5
six                           1.16.0
smmap                         5.0.1
sniffio                       1.3.0
sse-starlette                 1.8.2
starlette                     0.32.0.post1
streamlit                     1.29.0
sympy                         1.12
tenacity                      8.2.3
tiktoken                      0.5.2
tokenizers                    0.20.3
toml                          0.10.2
tomlkit                       0.12.0
toolz                         0.12.0
torch                         2.0.0+cu118
torchvision                   0.15.0+cu118
tornado                       6.4
tqdm                          4.67.1
transformers                  4.46.0
transformers-stream-generator 0.0.4
triton                        3.2.0
trl                           0.15.1
typer                         0.12.3
typing_extensions             4.9.0
tyro                          0.6.3
tzdata                        2024.2
tzlocal                       5.2
uc-micro-py                   1.0.3
urllib3                       2.1.0
uvicorn                       0.25.0
validators                    0.22.0
watchdog                      3.0.0
websockets                    11.0.3
wheel                         0.41.2
xxhash                        3.4.1
yarl                          1.9.4
zipp                          3.17.0
zstandard                     0.23.0`


reference code：
`# 3. 数据加载与格式转换
with open('data/RLFTqa_delete_some_answer_is_null.json', 'r') as f:
    data = json.load(f)
# 打乱数据顺序
random.shuffle(data)
data = data[0:200]
formatted_data = []
for item in data:
    formatted_data.append({
        "prompt": item["question"],
        "reference": "||".join(item.get("answer", []))
    })


# 定义数据集分割比例（70%训练，15%验证，15%测试）
n_total = len(formatted_data)
n_train = int(0.7 * n_total)
n_eval = int(0.15 * n_total)
n_test = n_total - n_train - n_eval

# 执行数据集分割
train_data = formatted_data[:1]
eval_data = formatted_data[n_train:n_train+n_eval]
# eval_data = formatted_data[n_train+n_eval:]
test_data = formatted_data[n_train+n_eval:]

# 定义统一的特征结构
dataset_features = Features({
    "prompt": Value("string"),
    "reference": Value("string")
})

# 创建三个数据集
train_dataset = Dataset.from_list(train_data, features=dataset_features)
eval_dataset = Dataset.from_list(eval_data, features=dataset_features)
test_dataset = Dataset.from_list(test_data, features=dataset_features)

# 4. 模型初始化配置优化
local_model_path = "./Qwen2.5-7B-Instruct-GPTQ-Int4"
tokenizer = AutoTokenizer.from_pretrained(local_model_path)
tokenizer.pad_token = tokenizer.eos_token

# 量化配置（根据实际硬件调整）
# quant_config = BitsAndBytesConfig(
#     load_in_4bit=True,
#     bnb_4bit_use_double_quant=True,
#     bnb_4bit_quant_type="nf4",
#     bnb_4bit_compute_dtype=torch.float16
# )

base_model = AutoModelForCausalLM.from_pretrained(
    local_model_path,
    # quantization_config=quant_config,
    device_map="auto",
    trust_remote_code=True,
    torch_dtype=torch.float16
)
base_model.config.gradient_checkpointing = False
# 5. LoRA配置优化
lora_config = LoraConfig(
    r=16,  # 适当增加秩维度
    lora_alpha=32,
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj"],  # 扩展目标模块
    lora_dropout=0.1,
    bias="none",
    task_type="CAUSAL_LM",
)

model = prepare_model_for_kbit_training(base_model)
model = get_peft_model(model, lora_config)

# 检查Peft模型参数
# print("可训练参数：")
# for name, param in model.named_parameters():
#     if param.requires_grad:
#         print(name)

# 6. GRPO配置参数优化
grpo_config = GRPOConfig(
    num_generations=8,  # 每组生成数量
    max_completion_length=512,  # 最大生成长度
    # batch_size=2,  # 设备级批次大小
    gradient_accumulation_steps=2,
    learning_rate=3e-5,
    # beta=0.02,  # KL散度系数

    # # 优化器配置
    # optim="adamw_torch",
    # weight_decay=0.01,

    # 日志与保存配置
    logging_steps=5,
    logging_first_step = True,
    save_steps=50,
    output_dir="./grpo_checkpoints",
    logging_dir="./grpo_checkpoints/log",
    log_level="info",
    log_completions=True,

    # 测试配置
    eval_strategy="steps",
    eval_steps=40,

    #测试配置
    do_predict=True,

    # 生成参数
    temperature=0.7,
    gradient_checkpointing=False,
    use_vllm=False


)
model.bfloat16()
# 7. 初始化GRPOTrainer（关键修改）
trainer = GRPOTrainer(
    model=model,
    args=grpo_config,
    reward_funcs=[reward_func, format_reward_func],  # 直接传递奖励函数
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    processing_class=tokenizer,
    # 关键参数说明：
    # reward_funcs接收自定义函数，函数参数会自动匹配数据集中的列（需要reference列）
    # processing_class必须显式传递tokenizer
)
train_output = trainer.predict(train_dataset)
# print(train_output)


# 8. 启动训练（自动处理生成和奖励计算）
trainer.train()

# 9. 保存最终适配器
model.save_pretrained("grpo_lora_adapter")`

As mentioned above, the length of train_det is 1. Before training, I first call trainer. predict (train_dataset). Then I printed the completion of the model in the library function, and the obtained location is as follows:
xxx/lib/python3.10/site-packages/trl/trainer/grpo_trainer.py：
`    def _prepare_inputs(self, inputs: dict[str, Union[torch.Tensor, Any]]) -> dict[str, Union[torch.Tensor, Any]]:
        device = self.accelerator.device
        prompts = [x["prompt"] for x in inputs]
        prompts_text = [maybe_apply_chat_template(example, self.processing_class)["prompt"] for example in inputs]
        prompt_inputs = self.processing_class(
            prompts_text, return_tensors="pt", padding=True, padding_side="left", add_special_tokens=False
        )
        prompt_inputs = super()._prepare_inputs(prompt_inputs)
        prompt_ids, prompt_mask = prompt_inputs["input_ids"], prompt_inputs["attention_mask"]
        if self.max_prompt_length is not None:
            prompt_ids = prompt_ids[:, -self.max_prompt_length :]
            prompt_mask = prompt_mask[:, -self.max_prompt_length :]

        # Generate completions using either vLLM or regular generation
        if self.args.use_vllm:
            print("Using VLM model")
            # First, have main process load weights if needed
            if self.state.global_step != self._last_loaded_step:
                self._move_model_to_vllm()
                self._last_loaded_step = self.state.global_step

            # Generate completions using vLLM: gather all prompts and use them in a single call in the main process
            all_prompts_text = gather_object(prompts_text)
            if self.accelerator.is_main_process:
                outputs = self.llm.generate(all_prompts_text, sampling_params=self.sampling_params, use_tqdm=False)
                completion_ids = [out.token_ids for completions in outputs for out in completions.outputs]
            else:
                completion_ids = [None] * len(all_prompts_text)
            # Broadcast the completions from the main process to all processes, ensuring each process receives its
            # corresponding slice.
            completion_ids = broadcast_object_list(completion_ids, from_process=0)
            process_slice = slice(
                self.accelerator.process_index * len(prompts),
                (self.accelerator.process_index + 1) * len(prompts),
            )
            completion_ids = completion_ids[process_slice]

            # Pad the completions, and concatenate them with the prompts
            completion_ids = [torch.tensor(ids, device=device) for ids in completion_ids]
            completion_ids = pad(completion_ids, padding_value=self.processing_class.pad_token_id)
            prompt_completion_ids = torch.cat([prompt_ids, completion_ids], dim=1)
        else:
            # Regular generation path
            with unwrap_model_for_generation(self.model, self.accelerator) as unwrapped_model:
                prompt_completion_ids = unwrapped_model.generate(
                    prompt_ids, attention_mask=prompt_mask, generation_config=self.generation_config
                )

            # Compute prompt length and extract completion ids
            prompt_length = prompt_ids.size(1)
            prompt_ids = prompt_completion_ids[:, :prompt_length]
            completion_ids = prompt_completion_ids[:, prompt_length:]

        # Mask everything after the first EOS token
        is_eos = completion_ids == self.processing_class.eos_token_id
        eos_idx = torch.full((is_eos.size(0),), is_eos.size(1), dtype=torch.long, device=device)
        eos_idx[is_eos.any(dim=1)] = is_eos.int().argmax(dim=1)[is_eos.any(dim=1)]
        sequence_indices = torch.arange(is_eos.size(1), device=device).expand(is_eos.size(0), -1)
        completion_mask = (sequence_indices <= eos_idx.unsqueeze(1)).int()

        # Concatenate prompt_mask with completion_mask for logit computation
        attention_mask = torch.cat([prompt_mask, completion_mask], dim=1)  # (B*G, P+C)

        logits_to_keep = completion_ids.size(1)  # we only need to compute the logits for the completion tokens

        with torch.inference_mode():
            if self.ref_model is not None:
                ref_per_token_logps = self._get_per_token_logps(
                    self.ref_model, prompt_completion_ids, attention_mask, logits_to_keep
                )
            else:
                with self.accelerator.unwrap_model(self.model).disable_adapter():
                    ref_per_token_logps = self._get_per_token_logps(
                        self.model, prompt_completion_ids, attention_mask, logits_to_keep
                    )

        # Decode the generated completions
        completions_text = self.processing_class.batch_decode(completion_ids, skip_special_tokens=True)
        print("Completions text:", completions_text)`

As I mentioned, when predicting, the model shows basic problem-solving ability, but during training, it doesn't even generate sentences that fit the task logic

**Comment by Wang-Xiaodong1899** [comment]
any update?

**Comment by Wang-Xiaodong1899** [comment]
I have also found that when the input prompt is too long, such as the length of my prompt input_id is 3964, I find that the complete result produces strange results. The normal inference model, on the other hand, is the normal output. Here I am using the model Qwen2-VL-7B.

strange completion such as `[<output>```python\n```]`, or `'<output>2003:000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000'`

Any suggestions?

**Comment by Wang-Xiaodong1899** [comment]
@qgallouedec

**Comment by skywalkerzhang** [comment]
any update? I found with the same input, the GRPOTrainer Produces Inferior Completions is very different from the prediction generated by model.generate(); inferior completion doesn't follow my instruction.

**Comment by CuriousG102** [comment]
Possible root cause: https://github.com/huggingface/trl/issues/3520

**Comment by albertvillanova** [comment]
Hey @yzhdut, sorry for the very slow reply — we're catching up on the backlog. 

The generation/tokenization code you quoted (the old `_prepare_inputs` that tokenized prompts with `add_special_tokens=False`) has been completely rewritten since 0.15.1: prompts are now tokenized via the chat template for conversational data and with the default `add_special_tokens=True` otherwise, so the BOS-stripping mechanism that #3520 flagged as a likely cause of garbage completions no longer exists on `main` (see https://github.com/huggingface/trl/blob/576a5f6bb536ca2fcfe934419d9f98f6f443eab8/trl/trainer/grpo_trainer.py#L1321-L1342). On the train-vs-`predict()` gap itself: rollouts during `train()` run with the model in training mode, so your `lora_dropout=0.1` is active during generation and will make completions noisier than the eval-mode ones `predict()` produces — that alone can explain a visible quality difference. Could you upgrade to the latest TRL and retest? If the gap is still there on a current version, please re-open with a minimal script and a fresh `trl env` and we'll dig in. 

Closing as outdated for now.
